src/t_hash.c

```diff
@@ -29,52 +29,183 @@
 
 #include "server.h"
 #include <math.h>
+#include <stdbool.h>
 
 /*-----------------------------------------------------------------------------
  * Hash Entry API
  *----------------------------------------------------------------------------*/
 
-struct hashTypeEntry {
-    sds value;
-    char field_offset;
-    char field_data[];
-};
+/* The hashTypeEntry pointer is the field sds. We encode the entry layout type
+ * in the field SDS header. Field type SDS_TYPE_5 doesn't have any spare bits to
+ * encode this so we use it only for the first layout type.
+ *
+ * Entry with embedded value, used for small sizes. The value is stored as
+ * SDS_TYPE_8. The field can use any SDS type.
+ *
+ *     +--------------+---------------+
+ *     | field        | value         |
+ *     | hdr "foo" \0 | hdr8 "bar" \0 |
+ *     +------^-------+---------------+
+ *            |
+ *            |
+ *          entry pointer = field sds
+ *
+ * Entry with value pointer, used for larger fields and values. The field is SDS
+ * type 8 or higher.
+ *
+ *     +-------+--------------+
+ *     | value | field        |
+ *     | ptr   | hdr "foo" \0 |
+ *     +-------+------^-------+
+ *                    |
+ *                    |
+ *                 entry pointer = field sds
+ */
+
+/* The maximum allocation size we want to use for entries with embedded
+ * values. */
+#define EMBED_VALUE_MAX_ALLOC_SIZE 128
+
+/* SDS aux flag. If set, it indicates that the entry has an embedded value
+ * pointer located in memory before the embedded field. If unset, the entry
+ * instead has an embedded value located after the embedded field. */
+#define FIELD_SDS_AUX_BIT_ENTRY_HAS_VALUE_PTR 0
+
+static inline bool entryHasValuePtr(const hashTypeEntry *entry) {
+    return sdsGetAuxBit(entry, FIELD_SDS_AUX_BIT_ENTRY_HAS_VALUE_PTR);
+}
+
+/* Returns the location of a pointer to a separately allocated value. Only for
+ * an entry without an embedded value. */
+static sds *hashTypeEntryGetValueRef(const hashTypeEntry *entry) {
+    serverAssert(entryHasValuePtr(entry));
+    char *field_data = sdsAllocPtr(entry);
+    field_data -= sizeof(sds *);
+    return (sds *)(void *)field_data;
```

Comment:
did not understand this....
```suggestion
    return (sds *)field_data;
```

Reply:
It's forbidden to cast from `char *` to `sds *`. They're not compatible pointer types. Only `void *` is compatible with all pointer types.

Reply:
mmm. O.K maybe some compilers will complain here (my clang did not). Go ahead.

Reply:
BTW - I think since field_data is (char *) than there is no issue here. sds is simply an alias for char* so I think there should be no issue.

Reply:
There is an issue. We want to return `sds *`, not `sds`.

`sds *` means `char **`.

~~I already got a compile warning here. That's why I know.~~

[Edit] Hm... maybe it was't here. Let me do some research about these casting rules...

Reply:
With just `return field_data;` and hoping for an implicit cast to `sds *` I get this error:

```
error: returning ‘char *’ from a function with incompatible return type ‘char **’ [-Wincompatible-pointer-types]
```

Reply:
well. I did suggest to return (sds*) field_data for that reason.